apps/labs/posts/sparse-array-ecosystem.md

```diff
@@ -0,0 +1,215 @@
+
+#### Dictionary-of-keys (DOK)
+
+This is one of the simplest formats to explain, so we begin with this one. The idea is that we have a dictionary that stores the coordinates of present values as the key, with the value representing the value of the array at that coordinate. Let's illustrate with a small example using the `sparse` library.
+
+```python
+>>> import numpy as np
+>>> import sparse
+>>> import pprint
+>>> eye_dok = sparse.DOK.from_numpy(np.eye(5))
+>>> pprint.pp(eye_dok.data)
+{(0, 0): 1.0, (1, 1): 1.0, (2, 2): 1.0, (3, 3): 1.0, (4, 4): 1.0}
+```
+
+This format has a number of advantages:
+
+1. It's simple to implement.
+2. It supports insertion into any coordinate by inserting the appropriate key-value pair into the dictionary.
+3. Random access is possible by retrieving a key from the dictionary.
+
+However, it also has a number of disadvantages:
+
+1. Iterating over values may be random, depending on the dictionary implementation. This creates problems when combining different arrays; and checking the presence of a given row or column (one can only check the presence of an element efficiently).
+2. It suffers from a higher memory usage than some other formats, as dictionaries take up more space than arrays.
+3. Retrieval of a single element may be fast, but it's slower than a similar access from an array due to hashing.
+
+### COOrdinate Format
+
+This format is also extremely simple, it stores one array of values, and a corresponding array of positions.
+
+```python
+>>> eye_coo = sparse.COO.from_numpy(np.eye(5))
+>>> eye_coo.coords
+array([[0, 1, 2, 3, 4],
+       [0, 1, 2, 3, 4]])
+>>> eye_coo.data
+array([1., 1., 1., 1., 1.])
+```
+
+The advantages for this one are the flip-side of DOK:
+
+1. It's simple to implement.
+2. If the constituent arrays are stored as lists, one can easily append a new coordinate. However, we risk working with out-of-order indices that way.
+3. Storage is near-optimal (for hyper-sparse arrays).
+4. Iterating in-order is possible if the constituent arrays are sorted.
+
+Here are some disadvantages:
+
+1. Random access using a coordinate isn't possible, one has to search for an existing coordiate. This may be easy if the coordinates are sorted or hard if they're not.
+2. Random insertion makes the coordinates go out of order.
+3. Sorting may be frequently required.
+
+## What tools can I use?
+
+### [`sparse`](https://sparse.pydata.org/)
+
+Also known as PyData/Sparse, this is a library maintained by yours truly. Its aim is to provide perfomant sparse array operations following the Array API standard, though we are some time away from reaching that goal. The work on this library started in April 2017, with the author of this blog post taking over maintenance around January 2018.
+
```

Comment:
To make these sections a bit meatier, maybe, for each library, demonstrate the same code snippet for each. The snippet could be a short script performing some basic operations.

Reply:
Okay. I see that the Array Creation section below basically achieves this ask. Maybe instead of having those examples below, you can interleave them here. For example, move L145-L160 to here. In the `scipy.sparse` section, move L164-L186 (splitting out the CuPy sparse example to the CuPy section). Similarly for PyTorch.

Then you can move the "Honorable Mentions" section to before `## API and Performance Comparison`.

That make sense?

Reply:
Done in https://github.com/Quansight/Quansight-website/pull/837/commits/4761fc3eb74286ade1d9344427b00551b44b5724.